python/pyspark/sql/udtf.py

```diff
@@ -133,12 +133,28 @@ class AnalyzeResult:
         If non-empty, this is a sequence of expressions that the UDTF is specifying for Catalyst to
         sort the input TABLE argument by. Note that the 'partitionBy' list must also be non-empty
         in this case.
+    acquireExecutionMemoryMbRequested: long
+        If this is not None, this represents the amount of memory in megabytes that the UDTF should
+        request from each Spark executor that it runs on. Then the UDTF takes responsibility to use
+        at most this much memory, including all allocated objects. The purpose of this functionality
+        is to prevent executors from crashing by running out of memory due to the extra memory
+        consumption invoked by the UDTF's 'eval' and 'terminate' and 'cleanup' methods. Spark will
+        then call 'TaskMemoryManager.acquireExecutionMemory' with the requested number of megabytes.
+    acquireExecutionMemoryMbActual: long
+        If there is a task context available, Spark will assign this field to the number of
+        megabytes returned from the call to the TaskMemoryManager.acquireExecutionMemory' method, as
+        consumed by the UDTF's'__init__' method. Therefore, its 'eval' and 'terminate' and 'cleanup'
+        methods will know it thereafter and can ensure to bound memory usage to at most this number.
+        Note that there is no effect if the UDTF's 'analyze' method assigns a value to this; it will
+        be overwritten.
     """
```

Comment:
@dtenedor I did some dumb memory profiling running the following UDTF

```python
from pyspark.sql.functions import udtf
import resource

@udtf(returnType="step: int, memory: int")
class SimpleUDTF:
    def __init__(self, *args, **kwargs):
        self.step_id = 0
    
    @staticmethod
    def get_peak_memory_usage_kb() -> int:
        return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

    def eval(self, *args, **kwargs):
        yield self.step_id, self.get_peak_memory_usage_kb()
        self.step_id += 1

    def terminate(self, *args, **kwargs):
        yield self.step_id, self.get_peak_memory_usage_kb()

spark.udtf.register("pyUdtfMemProfile", SimpleUDTF)
```

I'm not entirely confident in the results because I noticed running a high memory UDTF before the profiling UDTF would increase the peak memory usage (I guess they share a process?). But on a fresh 14.2 cluster I ran this and got max memory usage across 100 pyUdtfMemProfile invocations of `45440 KB`, which seems reasonable.

Is this memory already accounted for / pre-allocated on the executor? It's not clear to me if a user should request `processSize + additionalClassMemory` or just `additionalClassMemory` when they're specifying `acquireExecutionMemoryMbRequested` and `minMemoryMb`

Reply:
The `TaskMemoryManager.acquireExecutionMemory` API is a memory reservation system. The idea is that an operator that will consume memory in the future should call this method beforehand with the peak expected future memory usage on the executor. This comprises a memory reservation such that the sum of all such reservations may not reach or exceed the total memory available on the executor (less some fixed overhead for the executor's own data structures).

For the case of Python UDTFs, the function should set `AnalyzeResult. acquireExecutionMemoryMbRequested` to the max expected memory use of the python process, including memory allocated by the function itself as well as any imported libraries, etc.

Reply:
Makes sense! What do you think about this being difficult to set, especially for UDTF developers? If you think the test I did with `pyUdtfMemProfile` is a reasonable estimate, what do we think of setting a global `minMemoryMb` to something like 100MB? I think that could make the manual memory assignment less prone to user-error. It's probably also good to have a floor on some level that way the number of UDTFs simultaneously running on an executor has a hard ceiling

Reply:
Sounds good, I added a default of 100MB, which any UDTF can (and probably should) override to a more specific number.